src/nbl/asset/utils/CSmoothNormalGenerator.cpp

```diff
@@ -281,6 +337,208 @@ std::array<uint32_t, 8> CSmoothNormalGenerator::VertexHashMap::getNeighboringCel
+
+	core::vector<std::optional<uint32_t>> groupIndexes(vertices.getVertexCount());
+
+	auto canJoinVertices = [&](uint32_t index1, uint32_t index2)-> bool
+  {
+    if (!isAttributeEqual(polygon->getPositionView(), index1, index2, epsilon))
+			return false;
+    if (!isAttributeEqual(polygon->getNormalView(), index1, index2, epsilon))
+			return false;
+		for (const auto& jointWeightView : polygon->getJointWeightViews())
+		{
+			if (!isAttributeEqual(jointWeightView.indices, index1, index2, epsilon)) return false;
+			if (!isAttributeEqual(jointWeightView.weights, index1, index2, epsilon)) return false;
+		}
+		for (const auto& auxAttributeView : polygon->getAuxAttributeViews())
+			if (!isAttributeEqual(auxAttributeView, index1, index2, epsilon)) return false;
+
+		return true;
+  };
+
+	for (uint32_t cell = 0; cell < vertices.getBucketCount() - 1; cell++)
+	{
+		VertexHashMap::BucketBounds processedBucket = vertices.getBucketBoundsById(cell);
+
+		for (core::vector<CPolygonGeometryManipulator::SSNGVertexData>::iterator processedVertex = processedBucket.begin; processedVertex != processedBucket.end; processedVertex++)
+		{
+			std::array<uint32_t, 8> neighboringCells = vertices.getNeighboringCellHashes(*processedVertex);
+			hlsl::float32_t3 normal = processedVertex->parentTriangleFaceNormal * processedVertex->wage;
+
+			auto& groupIndex = groupIndexes[processedVertex->index];
+
+			//iterate among all neighboring cells
+			for (int i = 0; i < 8; i++)
+			{
+				VertexHashMap::BucketBounds bounds = vertices.getBucketBoundsByHash(neighboringCells[i]);
+				for (auto neighbourVertex_it = bounds.begin; neighbourVertex_it != bounds.end; neighbourVertex_it++)
+				{
+					const auto neighbourGroupIndex = groupIndexes[neighbourVertex_it->index];
+
+					hlsl::float32_t3 normal1, normal2;
+					polygon->getNormalView().decodeElement(processedVertex->index, normal1);
+					polygon->getNormalView().decodeElement(neighbourVertex_it->index, normal2);
+
+					hlsl::float32_t3 position1, position2;
+					polygon->getPositionView().decodeElement(processedVertex->index, position1);
+					polygon->getPositionView().decodeElement(neighbourVertex_it->index, position2);
+					 
+					// find the first group that this vertex can join
+					if (processedVertex != neighbourVertex_it && neighbourGroupIndex && canJoinVertices(processedVertex->index, neighbourVertex_it->index))
+					{
+						groupIndex = neighbourGroupIndex;
+						break;
+					}
+				}
+			}
+			if (!groupIndex)
+			{
+        // create new group if no group nearby that is compatible with this vertex
+        groupIndex = groups.size();
+        groups.push_back({ processedVertex->index});
+			}
+		}
+	}
+
+  auto outPolygon = core::move_and_static_cast<ICPUPolygonGeometry>(polygon->clone(0u));
+	outPolygon->setIndexing(IPolygonGeometryBase::TriangleList());
+
+	const uint32_t indexSize = (groups.size() < std::numeric_limits<uint16_t>::max()) ? sizeof(uint16_t) : sizeof(uint32_t);
```

Comment:
wrong thing to decide, while you iterate above keep track of the maxVertex value you actually use (so in new refactor `remappedVertexID` AFTER neighbours are visited) 

Then use that to determine if you need uint16, uint32 or uint64 index buffer on the output. 

**!!! Also right now you don't set the new and correct `maxVertex` (your attribute views don't get shortened) on the `outPolygon` !!!**

Reply:
if you want to demote the format, then you need to make a view with matching format as input, use that during the iteration, and then copy over to a newer view with narrower format